data/transactions/logic/eval.go

```diff
@@ -1013,6 +1095,10 @@ func EvalContract(program []byte, gi int, aid basics.AppIndex, params *EvalParam
 		txn:        &params.TxnGroup[gi],
 		appID:      aid,
 	}
+	// Save scratch for `gload`. We used to copy, but cx.scratch is quite large,
+	// about 8k, and caused measurable CPU and memory demands.  Of course, these
+	// should never be changed by later transactions.
+	cx.pastScratch[cx.groupIndex] = &cx.Scratch
```

Comment:
enforce with a test for opcodes accessing `cx.pastScratch` ?

Reply:
`func TestGload(t *testing.T)` ought to cover it.

This comment actually happened a while ago, it only appears here because I move it a bit. The only change in this PR is that `cx.pastScratch` is an array embedded in `EvalContext`, rather than slice, so we don't have to worry about allocating it. It simplified accessing the pastScratch when preparing the structured error. (It also should be faster - a slightly larger EvalContext, with one less individual allocation.)